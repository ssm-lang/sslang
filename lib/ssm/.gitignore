# Prerequisites
*.d

# Object files
*.o
*.ko
*.obj
*.elf
counter
counter2
clock
fib
fib2
fib3
onetwo

# Linker output
*.ilk
*.map
*.exp

# Precompiled Headers
*.gch
*.pch

# Libraries
*.lib
*.a
*.la
*.lo

# Shared objects (inc. Windows DLLs)
*.dll
*.so
*.so.*
*.dylib

# Executables
*.exe
*.app
*.i*86
*.x86_64
*.hex

# Debug files
*.dSYM/
*.su
*.idb
*.pdb

# Kernel Module Compile Results
*.mod*
*.cmd
.tmp_versions/
modules.order
Module.symvers
Mkfile.old
dkms.conf

# gcov output
*.gcov
*.gcda
*.gcno

build/

.pio

compile_commands.json

vgcore.*
