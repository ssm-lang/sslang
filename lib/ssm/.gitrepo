; DO NOT EDIT (unless you know what you are doing)
;
; This subdirectory is a git "subrepo", and this file is maintained by the
; git-subrepo command. See https://github.com/git-commands/git-subrepo#readme
;
[subrepo]
	remote = ssm-runtime
	branch = main
	commit = a9bf5272efed87994389613dd29e8f99bb7465c5
	parent = bfbde4f3b1727f9a6ffe768391f10c50e3f2fbdb
	method = merge
	cmdver = 0.4.3
