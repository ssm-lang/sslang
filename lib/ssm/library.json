{
    "name": "ssm",
    "version": "0.1",
    "description": "A lightweight runtime system for the sparse synchronous model, a deterministic real-time system model",
    "repository":
    {
	"type":"git",
	"url":"https://github.com/sedwards-lab/ssm.git"
    },
    "authors":
    [
	{"name":"Stephen A. Edwards",
	 "email":"sedwards@cs.columbia.edu",
	 "url":"http://www.cs.columbia.edu/~sedwards",
	 "maintainer": true
	}
    ],
    "license":"BSD-3-Clause",
    "frameworks":"*",
    "platform":"*"
}
    
