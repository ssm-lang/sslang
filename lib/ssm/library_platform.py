Import("env")
from os.path import join, realpath

platform = env['PIOFRAMEWORK'][0]
board = env['PIOENV']

print("hello world")
print("platform: ", platform)
print("board: ", board)

platform_dir = realpath(join("platform", platform))
boards_dir = realpath(join(platform_dir, "boards"))
board_dir = realpath(join(boards_dir, board))

# Specify platform/board includes
env.Append(
    CPPPATH=[
        realpath(join("platform", platform, "include")),
        realpath(join("platform", platform, "boards", board, "include")),
    ]
)
# Specify common and platform source directories
env.Replace(
    SRC_FILTER=[
        "-<*>",
        "+<src>",
        "+<%s>" % realpath(join("platform", platform, "src")),
        "+<%s>" % realpath(join("platform", platform, "boards", board, "src")),
    ]
)
# Platform defines
env.Append(
    CPPDEFINES=[
      # This will be needed if we ever have a 64-bit timer platform,
      # to set PLATFORM_TIMER64; e.g.:
      # "PLATFORM_TIMER64",
    ]
)
