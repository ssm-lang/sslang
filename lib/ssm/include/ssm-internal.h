/** @file ssm-internal.h
 *  @brief The internal interface of the SSM runtime.
 *
 *  This header file contains definitions and declarations that should not be
 *  exposed to user code.
 *
 *  @author Stephen Edwards (sedwards-lab)
 *  @author John Hui (j-hui)
 */
#ifndef _SSM_SCHED_H
#define _SSM_SCHED_H

#include <ssm.h>

/** @ingroup error
 *  @brief Throw an internal error.
 *
 *  @platformonly
 *
 *  @param cond the condition to assert.
 */
#define SSM_ASSERT(cond)                                                       \
  do                                                                           \
    if (!(cond))                                                               \
      SSM_THROW(SSM_INTERNAL_ERROR);                                           \
  while (0)

/** @ingroup time
 *  @brief The time of the next event in the event queue.
 *
 *  Used to determine whether and when to invoke ssm_tick().
 *
 *  @platformonly
 *
 *  @returns the next event time, or #SSM_NEVER if the event queue is empty.
 */
ssm_time_t ssm_next_event_time(void);

/** @ingroup time
 *  @brief Reset the scheduler.
 *
 *  Set #now to 0; clear the event and activation record queues.
 *
 *  This does not need to be called before calling ssm_tick() for the first
 *  time; the global state automatically starts initialized.
 *
 *  @platformonly
 */
void ssm_reset(void);

/** @ingroup time
 *  @brief Advance the current model time.
 *
 *  @a next must be later than ssm_now(), and earlier than or equal to
 *  ssm_next_event_time().
 *
 *  Exposed so that platform code can perform external variable updates, to
 *  implement external inputs.
 *
 *  @platformonly
 *
 *  @param next the time to advance to.
 *
 *  @throws SSM_INVALID_TIME  @a next is earlier than or equal to #now.
 *  @throws SSM_NOT_READY     @a next is later than the earliest queued event.
 */
void ssm_set_now(ssm_time_t next);

/** @ingroup sv
 *  @brief Perform a (delayed) update on a variable.
 *
 *  Schedules all routine activation records sensitive to @a sv in the
 *  activation queue.
 *
 *  Should only be called if the variable is scheduled to be updated #now.
 *
 *  Exposed so that platform code can perform external variable updates.
 *
 *  @platformonly
 *
 *  @param sv the variable.
 *
 *  @throws SSM_NOT_READY   @a sv was not scheduled to be updated #now.
 */
void ssm_update(ssm_sv_t *sv);

/** @ingroup sv
 *  @brief Unschedule any pending events on a variable.
 *
 *  Should be called before the variable is dropped.
 *
 *  Nothing happens if the variable does not have a pending event.
 *
 *  @platformonly
 *
 *  @param var  the variable.
 */
void ssm_unschedule(ssm_sv_t *var);

/** @ingroup act
 *  @brief Run the system for the next scheduled instant.
 *
 *  If there is nothing left to run in the current instant, advance #now to the
 *  time of the earliest event in the queue, if any.
 *
 *  Removes every event at the head of the event queue scheduled for #now,
 *  updates each variable's current value from its later value, and schedules
 *  all sensitive triggers in the activation queue. Finally, execute the step
 *  functions of all scheduled routines, in order of priority.
 *
 *  Should only be called if there are activation records to execute #now,
 *  or if #now is earlier than ssm_next_event_time().
 *
 *  @platformonly
 *
 *  @throws SSM_INTERNAL_ERROR  there are stale events in the queue before #now.
 */
void ssm_tick(void);

/** @ingroup adt
 *  @brief Compute the size of a heap-allocated ADT.
 *  @todo document
 */
#define ssm_adt_size(vc) (sizeof(struct ssm_adt1) + sizeof(ssm_value_t) * ((vc) - 1))

/**
 * @addtogroup mem
 * @{
 */

/** @brief The different kinds of heap objects, enumerated.
 *
 *  Types enumerated here that are not ADTs are chosen because they cannot be
 *  easily or efficiently expressed as a product of words. For instance, 64-bit
 *  timestamps cannot be directly stored in the payload of a regular heap
 *  object, where even-numbered timestamps may be misinterpreted as pointers.
 */
enum ssm_kind {
  SSM_ADT_K = 0,  /**< ADT object, e.g., #ssm_adt1 */
  SSM_TIME_K,     /**< 64-bit timestamps, #ssm_time_t */
  SSM_SV_K,       /**< Scheduled variables, #ssm_sv_t */
};

/** @brief Initializes the underlying allocator system.
 *
 *  Memory pages are requested from the platform/OS on-demand via the provided
 *  @a alloc_page_handler. This handler must return a pointer to the beginning
 *  of a range of memory of size #SSM_MEM_PAGE_SIZE, and the page @em must be
 *  zeroed out. These pages are assumed to never be freed.
 *
 *  To support arbitrarily large allocations, SSM's allocator uses @a
 *  alloc_mem_handler to allocate memory, and @a free_mem_handler to release it.
 *  There are no requirements on the contents of memory allocated with @a
 *  alloc_mem_handler. These handlers may also assume they will not be invoked
 *  to request memory ranges of less than #SSM_MEM_POOL_MAX bytes.
 *
 *  @platformonly
 *
 *  @param alloc_page_handler   allocates pages.
 *  @param alloc_mem_handler    allocates arbitrarily large.
 *  @param free_mem_handler     frees pages allocated with @a alloc_mem_handler.
 */
void ssm_mem_init(void *(*alloc_page_handler)(void),
                  void *(*alloc_mem_handler)(size_t),
                  void (*free_mem_handler)(void *, size_t));

#ifndef SSM_MEM_POOL_MIN
/** @brief Block size of the smallest memory pool.
 *
 *  Must be strictly greater than the word size, i.e., the size of #ssm_word_t.
 *
 *  @platformonly
 */
#define SSM_MEM_POOL_MIN 16
#endif

#if SSM_MEM_POOL_MIN < SSM_POINTER_SIZE
#error SSM_MEM_POOL_MIN must be larger than word size.
#endif

#ifndef SSM_MEM_POOL_FACTOR_BASE2
/** @brief Factor between each successive memory pool size, in base 2.
 *
 *  Must be strictly greater than 0.
 *
 *  @platformonly
 */
#define SSM_MEM_POOL_FACTOR_BASE2 2
#endif

#if SSM_MEM_POOL_FACTOR_BASE2 < 1
#error SSM_MEM_POOL_FACTOR_BASE2 must be strictly greater than 0.
#endif

#ifndef SSM_MEM_POOL_COUNT
/** @brief Number of memory pools.
 *
 *  Must be strictly greater than 0.
 *
 *  @platformonly
 */
#define SSM_MEM_POOL_COUNT 4
#endif

#if SSM_MEM_POOL_COUNT < 1
#error SSM_MEM_POOL_COUNT must be strictly greater than 0.
#endif

/** @brief Compute the size of a memory pool.
 *
 *  @platformonly
 *
 *  @param pool   the 0-index of the memory pool.
 *  @returns      the size of the memory pool at position @a pool.
 */
#define SSM_MEM_POOL_SIZE(pool)                                                \
  (SSM_MEM_POOL_MIN << (SSM_MEM_POOL_FACTOR_BASE2 * pool))

/** @brief The size of the largest memory pool.
 *
 *  @platformonly
 */
#define SSM_MEM_POOL_MAX SSM_MEM_POOL_SIZE(SSM_MEM_POOL_COUNT - 1)

#ifndef SSM_MEM_PAGE_SIZE
/** @brief The size of a memory page; must be greater than #SSM_MEM_POOL_MAX.
 *
 *  @platformonly
 */
#define SSM_MEM_PAGE_SIZE SSM_MEM_POOL_SIZE(SSM_MEM_POOL_COUNT)
#endif

/** @} */

#endif /* _SSM_SCHED_H */
