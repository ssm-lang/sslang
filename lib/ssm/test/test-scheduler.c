#include <ssm-internal.h>
#include <ssm.h>
#include <stdio.h>
#include <string.h>

#ifndef SSM_DEBUG
#error "SSM_DEBUG undefined; it should be defined for the library"
#endif

#define DUMMY_VALUE ssm_marshal(0)

typedef uint16_t q_idx_t;
extern ssm_sv_t *event_queue[];
extern q_idx_t event_queue_len;
extern void event_queue_consistency_check(void);
extern void event_queue_percolate_down(q_idx_t, ssm_sv_t *);

extern ssm_act_t *act_queue[];
extern q_idx_t act_queue_len;
extern void act_queue_consistency_check(void);
extern void act_queue_percolate_down(q_idx_t, ssm_act_t *);

#define NUM_VARIABLES 1024
ssm_value_t variables[NUM_VARIABLES];

#define NUM_ACTS 1024
ssm_act_t acts[NUM_ACTS];

#define NUM_TRIGGERS 1024
ssm_trigger_t triggers[NUM_TRIGGERS];

void check_starts_initialized() {
  SSM_ASSERT(ssm_now() == 0L);
  SSM_ASSERT(event_queue_len == 0);
  SSM_ASSERT(act_queue_len == 0);
}

void reset_all() {
  for (int i = 0; i < NUM_VARIABLES; i++)
    ssm_drop(variables[i]);
  ssm_reset();
  for (int i = 0; i < NUM_VARIABLES; i++)
    variables[i] = ssm_new_sv(DUMMY_VALUE);

  check_starts_initialized();
}

void event_queue_basic() {
  reset_all();
  SSM_ASSERT(ssm_next_event_time() == SSM_NEVER);
  SSM_ASSERT(ssm_to_sv(variables[0])->last_updated != ssm_now());
  ssm_later(variables[0], 1, DUMMY_VALUE);
  SSM_ASSERT(event_queue_len == 1);
  SSM_ASSERT(ssm_next_event_time() == 1);
  event_queue_consistency_check();
  ssm_tick();
  SSM_ASSERT(ssm_to_sv(variables[0])->last_updated == ssm_now());
  SSM_ASSERT(ssm_now() == 1);
  SSM_ASSERT(ssm_next_event_time() == SSM_NEVER);
  SSM_ASSERT(event_queue_len == 0);
}

void print_event_queue() {
  printf("Event queue: ");
  for (int i = 1; i <= event_queue_len; i++)
    if (event_queue[i])
      printf("%c", (char)event_queue[i]->later_time);
    else
      printf("[NULL]");
  printf("\n");
}

/*** Fill the event queue with events whose times are the characters
 * in the input string, schedule each, then repeatedly remove the
 * earliest from the queue, making sure they come out in the order
 * given by the expected string.
 */
void event_queue_sort_string(const char *input, const char *expected) {
  reset_all();
  SSM_ASSERT(event_queue_len == 0);
  ssm_value_t *var = variables;
  for (const char *cp = input; *cp; ++cp, ++var) {
    ssm_later(*var, (ssm_time_t)*cp, DUMMY_VALUE);
    event_queue_consistency_check();
  }

  while (event_queue_len) {
    char c = (char)event_queue[1]->later_time;
    printf("%c", c);
    SSM_ASSERT(c == *expected++);
    event_queue[1]->later_time = SSM_NEVER;
    ssm_sv_t *to_insert = event_queue[event_queue_len--]; // get last

    if (event_queue_len) // Was this the last?
      event_queue_percolate_down(1, to_insert);
    event_queue_consistency_check();
  }
  printf("\n");
}

/** Schedule the characters of a string as events whose times are the
 * characters, then reschedule each, replacing non-space characters with
 * lowercase versions of each, then run ssm_tick() until the event queue is
 * empty.  The sequence of output characters is a sorted, unique list of
 * all the characters in the string"
 *
 */
void event_queue_reschedule_string(const char *input, const char *expected) {
  reset_all();
  ssm_value_t *var = variables;

  for (const char *cp = input; *cp; ++cp, ++var) {
    ssm_later(*var, (ssm_time_t)*cp, DUMMY_VALUE);
    event_queue_consistency_check();
  }

  // print_event_queue();

  // Reschedule each element, swapping its case

  var = variables;
  for (const char *cp = input; *cp; ++cp, ++var) {
    if (*cp != ' ') {
      if (*cp > 'Z')
        ssm_later(*var, (ssm_time_t)*cp + 'A' - 'a', DUMMY_VALUE);
      else
        ssm_later(*var, (ssm_time_t)*cp + 'a' - 'A', DUMMY_VALUE);
    }
    event_queue_consistency_check();
  }

  // print_event_queue();

  while (ssm_next_event_time() != SSM_NEVER) {
    ssm_tick();
    event_queue_consistency_check();
    char c = ssm_now();
    printf("%c", c);
    SSM_ASSERT(c == *expected++);
  }
  printf("\n");
}

/*** Fill the event queue with events times by the characters in the
 * string, then unschedule the first n and verify what's left
 */
void event_queue_unschedule_string(const char *input, int n,
                                   const char *expected) {
  reset_all();
  ssm_value_t *var = variables;
  for (const char *cp = input; *cp; ++cp, ++var) {
    ssm_later(*var, (ssm_time_t)*cp, DUMMY_VALUE);
    event_queue_consistency_check();
  }

  var = variables;
  for (int j = n; j; ++var, --j) {
    ssm_unschedule(ssm_to_sv(*var));
    event_queue_consistency_check();
  }

  // Unschedule them again; it shouldn't matter

  var = variables;
  for (int j = n; j; ++var, --j) {
    ssm_unschedule(ssm_to_sv(*var));
    event_queue_consistency_check();
  }

  while (event_queue_len) {
    char c = (char)event_queue[1]->later_time;
    printf("%c", c);
    SSM_ASSERT(c == *expected++);
    event_queue[1]->later_time = SSM_NEVER;
    ssm_sv_t *to_insert = event_queue[event_queue_len--]; // get last

    if (event_queue_len) // Was this the last?
      event_queue_percolate_down(1, to_insert);
    event_queue_consistency_check();
  }
  printf("\n");
}

void act_queue_basic() {
  reset_all();
  SSM_ASSERT(act_queue_len == 0);
  SSM_ASSERT(!acts[0].scheduled);
  ssm_activate(acts);
  SSM_ASSERT(act_queue_len == 1);
  act_queue_consistency_check();
  ssm_tick();
  SSM_ASSERT(act_queue_len == 0);
}

void act_queue_sort_string(const char *input, const char *expected) {
  reset_all();
  SSM_ASSERT(act_queue_len == 0);
  ssm_act_t *act = acts;
  for (const char *cp = input; *cp; ++cp, ++act) {
    act->scheduled = false;
    act->priority = *cp;
    ssm_activate(act);
    act_queue_consistency_check();
  }

  while (act_queue_len) {
    char c = (char)act_queue[1]->priority;
    printf("%c", c);
    SSM_ASSERT(c == *expected++);
    ssm_act_t *to_insert = act_queue[act_queue_len--]; // get last

    if (act_queue_len) // Was this the last?
      act_queue_percolate_down(1, to_insert);
    act_queue_consistency_check();
  }
  printf("\n");
}

const char *next_expected;

void check_priority_step(ssm_act_t *act) {
  SSM_ASSERT(act);
  char c = act->priority;
  printf("%c", c);
  SSM_ASSERT(c == *next_expected++);
}

/** Enter the string into the activation record queue with priorities
 * then run ssm_tick() making the "step" function print and check the
 * priority of the activation record just executed.
 */
void act_queue_sort_tick(const char *input, const char *expected) {
  reset_all();
  SSM_ASSERT(act_queue_len == 0);
  ssm_act_t *act = acts;
  for (const char *cp = input; *cp; ++cp, ++act) {
    act->step = check_priority_step;
    act->scheduled = false;
    act->priority = *cp;
    ssm_activate(act);
    SSM_ASSERT(act->scheduled);
    act_queue_consistency_check();
  }

  next_expected = expected;

  ssm_tick();
  act_queue_consistency_check();

  SSM_ASSERT(*next_expected ==
             0); // Did we end up at the end of the expected string?

  SSM_ASSERT(act_queue_len ==
             0); // Should have emptied the activation record queue
  printf("\n");

  // Make sure all the activation records were unscheduled
  act = acts;
  for (const char *cp = input; *cp; ++cp, ++act)
    SSM_ASSERT(!act->scheduled);
}

bool step0ran, step1ran;

void step0(ssm_act_t *act) {
  step0ran = true;
  printf("step0 ");
}
void step1(ssm_act_t *act) {
  step1ran = true;
  printf("step1 ");
}

void trigger_basic() {
  reset_all();

  acts[0].step = step0;
  acts[0].priority = 3;
  SSM_ASSERT(!acts[0].scheduled);
  acts[1].step = step1;
  acts[1].priority = 4;
  SSM_ASSERT(!acts[1].scheduled);

  triggers[0].act = &acts[0];
  triggers[1].act = &acts[1];
  triggers[2].act = &acts[0];

  step0ran = step1ran = false;

  ssm_sensitize(variables[0], &triggers[0]);
  ssm_sensitize(variables[0], &triggers[1]);
  ssm_sensitize(variables[0], &triggers[2]);

  ssm_later(variables[0], 1, DUMMY_VALUE);

  ssm_tick();
  printf("\n");
  SSM_ASSERT(ssm_now() == 1);

  SSM_ASSERT(step0ran);
  SSM_ASSERT(step1ran);

  SSM_ASSERT(ssm_next_event_time() == SSM_NEVER);

  ssm_later(variables[0], 2, DUMMY_VALUE);

  step0ran = step1ran = false;
  ssm_tick();
  printf("\n");

  SSM_ASSERT(ssm_now() == 2);
  SSM_ASSERT(step0ran);
  SSM_ASSERT(step1ran);

  SSM_ASSERT(ssm_next_event_time() == SSM_NEVER);

  ssm_later(variables[0], 3, DUMMY_VALUE);
  ssm_desensitize(&triggers[1]);

  step0ran = step1ran = false;
  ssm_tick();
  printf("\n");

  SSM_ASSERT(ssm_now() == 3);
  SSM_ASSERT(step0ran);
  SSM_ASSERT(!step1ran);

  SSM_ASSERT(ssm_next_event_time() == SSM_NEVER);

  ssm_later(variables[0], 4, DUMMY_VALUE);
  ssm_desensitize(&triggers[0]);

  step0ran = step1ran = false;
  ssm_tick();
  printf("\n");

  SSM_ASSERT(ssm_now() == 4);
  SSM_ASSERT(step0ran);
  SSM_ASSERT(!step1ran);

  SSM_ASSERT(ssm_next_event_time() == SSM_NEVER);

  ssm_later(variables[0], 5, DUMMY_VALUE);
  ssm_desensitize(&triggers[2]);

  step0ran = step1ran = false;
  ssm_tick();
  printf("\n");

  SSM_ASSERT(ssm_now() == 5);
  SSM_ASSERT(!step0ran);
  SSM_ASSERT(!step1ran);

  SSM_ASSERT(ssm_next_event_time() == SSM_NEVER);

  ssm_later(variables[0], 6, DUMMY_VALUE);
  ssm_sensitize(variables[0], &triggers[1]);
  ssm_sensitize(variables[0], &triggers[0]);

  step0ran = step1ran = false;
  ssm_tick();
  printf("\n");

  SSM_ASSERT(ssm_now() == 6);
  SSM_ASSERT(step0ran);
  SSM_ASSERT(step1ran);

  SSM_ASSERT(ssm_next_event_time() == SSM_NEVER);

  ssm_desensitize(&triggers[0]);
  ssm_later(variables[0], 7, DUMMY_VALUE);

  step0ran = step1ran = false;
  ssm_tick();
  printf("\n");

  SSM_ASSERT(ssm_now() == 7);
  SSM_ASSERT(!step0ran);
  SSM_ASSERT(step1ran);

  SSM_ASSERT(ssm_next_event_time() == SSM_NEVER);

  ssm_sensitize(variables[0], &triggers[0]);

  for (ssm_trigger_t *trig = ssm_to_sv(variables[0])->triggers; trig;
       trig = trig->next)
    if (trig->act->priority > 0)
      ssm_activate(trig->act);
  step0ran = step1ran = false;
  ssm_tick();
  printf("\n");

  SSM_ASSERT(ssm_now() == 7);
  SSM_ASSERT(step0ran);
  SSM_ASSERT(step1ran);

  for (ssm_trigger_t *trig = ssm_to_sv(variables[0])->triggers; trig;
       trig = trig->next)
    if (trig->act->priority > 3)
      ssm_activate(trig->act);

  step0ran = step1ran = false;
  ssm_tick();
  printf("\n");

  SSM_ASSERT(ssm_now() == 7);
  SSM_ASSERT(!step0ran);
  SSM_ASSERT(step1ran);
}

void vacuous_step(ssm_act_t *act) {}

void ssm_throw(enum ssm_error reason, const char *file, int line,
               const char *func) {
  printf("SSM error at %s:%s:%d: reason: %d\n", file, func, line, reason);
  exit(1);
}

#define MAX_PAGES 2048
static void *pages[MAX_PAGES];
static size_t allocated_pages = 0;

static void *alloc_page(void) {
  if (allocated_pages >= MAX_PAGES) {
    SSM_THROW(SSM_EXHAUSTED_MEMORY);
    exit(3);
  }
  void *m = pages[allocated_pages++] = malloc(SSM_MEM_PAGE_SIZE);
  memset(m, 0, SSM_MEM_PAGE_SIZE);
  return m;
}

static void *alloc_mem(size_t size) { return malloc(size); }

static void free_mem(void *mem, size_t size) { free(mem); }

int main(void) {
  ssm_mem_init(alloc_page, alloc_mem, free_mem);

  for (int i = 0; i < NUM_VARIABLES; i++)
    variables[i] = ssm_new_sv(DUMMY_VALUE);

  for (int i = 0; i < NUM_ACTS; i++)
    acts[i] = (ssm_act_t){
        .step = vacuous_step,
        .caller = &ssm_top_parent,
        .pc = 0,
        .children = 0,
        .priority = 0,
        .depth = 0,
        .scheduled = false,
    };

  check_starts_initialized();

  event_queue_basic();
  event_queue_sort_string("", "");
  event_queue_sort_string("ABC", "ABC");
  event_queue_sort_string("CBA", "ABC");
  event_queue_sort_string("ABCD", "ABCD");
  event_queue_sort_string("DBCA", "ABCD");
  event_queue_sort_string("DCBA", "ABCD");
  event_queue_sort_string("MR JOCK TV QUIZ PHD BAGS FEW LYNX",
                          "       ABCDEFGHIJKLMNOPQRSTUVWXYZ");
  event_queue_sort_string("WALTZ BAD NYMPH FOR QUICK JIGS VEX",
                          "      AABCDEFGHIIJKLMNOPQRSTUVWXYZ");
  event_queue_sort_string("SPHINX OF BLACK QUARTZ JUDGE MY VOW",
                          "      AABCDEFGHIJKLMNOOPQRSTUUVWXYZ");
  event_queue_sort_string("THE QUICK BROWN FOX JUMPS OVER THE LAZY DOG",
                          "        ABCDEEEFGHHIJKLMNOOOOPQRRSTTUUVWXYZ");
  event_queue_sort_string("The Quick Brown Fox Jumps Over The Lazy Dog",
                          "        BDFJLOQTTaceeeghhikmnoooprrsuuvwxyz");

  event_queue_reschedule_string("ABC", "abc");
  event_queue_reschedule_string("CBA", "abc");
  event_queue_reschedule_string("ABCD", "abcd");
  event_queue_reschedule_string("DBCA", "abcd");
  event_queue_reschedule_string("DCBA", "abcd");
  event_queue_reschedule_string("ABBA SAID GO AWAY", " abdgioswy");
  event_queue_reschedule_string("MR JOCK TV QUIZ PHD BAGS FEW LYNX",
                                " abcdefghijklmnopqrstuvwxyz");
  event_queue_reschedule_string("WALTZ BAD NYMPH FOR QUICK JIGS VEX",
                                " abcdefghijklmnopqrstuvwxyz");
  event_queue_reschedule_string("SPHINX OF BLACK QUARTZ JUDGE MY VOW",
                                " abcdefghijklmnopqrstuvwxyz");
  event_queue_reschedule_string("THE QUICK BROWN FOX JUMPS OVER THE LAZY DOG",
                                " abcdefghijklmnopqrstuvwxyz");
  event_queue_reschedule_string("The Quick Brown Fox Jumps Over The Lazy Dog",
                                " ACEGHIKMNOPRSUVWXYZbdfjloqt");

  event_queue_unschedule_string("ABCD", 1, "BCD");
  event_queue_unschedule_string("ABCD", 2, "CD");
  event_queue_unschedule_string("ABCD", 3, "D");
  event_queue_unschedule_string("bcda", 2, "ad");

  event_queue_unschedule_string("The Quick Brown Fox Jumps Over The Lazy Dog",
                                4, "       BDFJLOQTaceeghikmnoooprrsuuvwxyz");
  event_queue_unschedule_string("The Quick Brown Fox Jumps Over The Lazy Dog",
                                15, "      DFJLOTaeeghmooprsuvxyz");

  act_queue_basic();

  act_queue_sort_string("", "");
  act_queue_sort_string("ABC", "ABC");
  act_queue_sort_string("CBA", "ABC");
  act_queue_sort_string("ABCD", "ABCD");
  act_queue_sort_string("DBCA", "ABCD");
  act_queue_sort_string("DCBA", "ABCD");
  act_queue_sort_string("MR JOCK TV QUIZ PHD BAGS FEW LYNX",
                        "       ABCDEFGHIJKLMNOPQRSTUVWXYZ");
  act_queue_sort_string("WALTZ BAD NYMPH FOR QUICK JIGS VEX",
                        "      AABCDEFGHIIJKLMNOPQRSTUVWXYZ");
  act_queue_sort_string("SPHINX OF BLACK QUARTZ JUDGE MY VOW",
                        "      AABCDEFGHIJKLMNOOPQRSTUUVWXYZ");
  act_queue_sort_string("THE QUICK BROWN FOX JUMPS OVER THE LAZY DOG",
                        "        ABCDEEEFGHHIJKLMNOOOOPQRRSTTUUVWXYZ");
  act_queue_sort_string("The Quick Brown Fox Jumps Over The Lazy Dog",
                        "        BDFJLOQTTaceeeghhikmnoooprrsuuvwxyz");

  act_queue_sort_tick("", "");
  act_queue_sort_tick("ABC", "ABC");
  act_queue_sort_tick("CBA", "ABC");
  act_queue_sort_tick("ABCD", "ABCD");
  act_queue_sort_tick("DBCA", "ABCD");
  act_queue_sort_tick("DCBA", "ABCD");

  act_queue_sort_tick("MR JOCKTVQUIZPHDBAGSFEWLYNX",
                      " ABCDEFGHIJKLMNOPQRSTUVWXYZ");
  act_queue_sort_tick("MR JOCK TV QUIZ PHD BAGS FEW LYNX",
                      "       ABCDEFGHIJKLMNOPQRSTUVWXYZ");
  act_queue_sort_tick("WALTZ BAD NYMPH FOR QUICK JIGS VEX",
                      "      AABCDEFGHIIJKLMNOPQRSTUVWXYZ");
  act_queue_sort_tick("SPHINX OF BLACK QUARTZ JUDGE MY VOW",
                      "      AABCDEFGHIJKLMNOOPQRSTUUVWXYZ");
  act_queue_sort_tick("THE QUICK BROWN FOX JUMPS OVER THE LAZY DOG",
                      "        ABCDEEEFGHHIJKLMNOOOOPQRRSTTUUVWXYZ");
  act_queue_sort_tick("The Quick Brown Fox Jumps Over The Lazy Dog",
                      "        BDFJLOQTTaceeeghhikmnoooprrsuuvwxyz");

  trigger_basic();

  for (int i = 0; i < NUM_VARIABLES; i++)
    ssm_drop(variables[i]);

  printf("PASSED\n");
  return 0;
}
