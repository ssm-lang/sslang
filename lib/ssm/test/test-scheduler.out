./build/test_test-scheduler
# 
# ABC
# ABC
# ABCD
# ABCD
# ABCD
#        ABCDEFGHIJKLMNOPQRSTUVWXYZ
#       AABCDEFGHIIJKLMNOPQRSTUVWXYZ
#       AABCDEFGHIJKLMNOOPQRSTUUVWXYZ
#         ABCDEEEFGHHIJKLMNOOOOPQRRSTTUUVWXYZ
#         BDFJLOQTTaceeeghhikmnoooprrsuuvwxyz
# abc
# abc
# abcd
# abcd
# abcd
#  abdgioswy
#  abcdefghijklmnopqrstuvwxyz
#  abcdefghijklmnopqrstuvwxyz
#  abcdefghijklmnopqrstuvwxyz
#  abcdefghijklmnopqrstuvwxyz
#  ACEGHIKMNOPRSUVWXYZbdfjloqt
# BCD
# CD
# D
# ad
#        BDFJLOQTaceeghikmnoooprrsuuvwxyz
#       DFJLOTaeeghmooprsuvxyz
# 
# ABC
# ABC
# ABCD
# ABCD
# ABCD
#        ABCDEFGHIJKLMNOPQRSTUVWXYZ
#       AABCDEFGHIIJKLMNOPQRSTUVWXYZ
#       AABCDEFGHIJKLMNOOPQRSTUUVWXYZ
#         ABCDEEEFGHHIJKLMNOOOOPQRRSTTUUVWXYZ
#         BDFJLOQTTaceeeghhikmnoooprrsuuvwxyz
# 
# ABC
# ABC
# ABCD
# ABCD
# ABCD
#  ABCDEFGHIJKLMNOPQRSTUVWXYZ
#        ABCDEFGHIJKLMNOPQRSTUVWXYZ
#       AABCDEFGHIIJKLMNOPQRSTUVWXYZ
#       AABCDEFGHIJKLMNOOPQRSTUUVWXYZ
#         ABCDEEEFGHHIJKLMNOOOOPQRRSTTUUVWXYZ
#         BDFJLOQTTaceeeghhikmnoooprrsuuvwxyz
# step0 step1 
# step0 step1 
# step0 
# step0 
# 
# step0 step1 
# step1 
# step0 step1 
# step1 
# PASSED
